Include/Rover/Range.hpp

```diff
@@ -68,121 +28,138 @@ namespace Rover {
+  Range(BeginFwd&&, EndFwd&&, GranularityFwd&&, Interval = Interval::CLOSED) ->
+    Range<std::decay_t<BeginFwd>, std::decay_t<EndFwd>,
+    std::decay_t<GranularityFwd>>;
 
-  template<typename Begin, typename End, typename Distribution,
-      typename Engine>
+  template<typename B, typename E, typename G>
   template<typename BeginFwd, typename EndFwd>
-  Range<Begin, End, Distribution, Engine>::Range
-      (BeginFwd&& begin, EndFwd&& end, Interval interval)
-      : m_engine(std::random_device()()),
-        m_begin(std::forward<BeginFwd>(begin)),
+  Range<B, E, G>::Range(BeginFwd&& begin, EndFwd&& end,
+      Interval interval)
+      : m_begin(std::forward<BeginFwd>(begin)),
         m_end(std::forward<EndFwd>(end)),
-        m_interval(interval) {}
-
-  template<typename Begin, typename End, typename Distribution,
-      typename Engine>
-  Range<Begin, End, Distribution, Engine>::Range
-      (Range<Begin, End, Distribution, Engine>&& other)
-      : m_begin(std::move(other.m_begin)),
-        m_end(std::move(other.m_end)),
-        m_interval(std::move(other.m_interval)) {}
-
-  template<typename Begin, typename End, typename Distribution,
-      typename Engine>
-  template<typename Session>
-  constexpr typename Range<Begin, End, Distribution, Engine>::Type
-      Range<Begin, End, Distribution, Engine>::generate(Session& session) {
+        m_round([](const Type& value){
+          return value;
+        }),
+        m_interval(interval),
+        m_engine(std::random_device()()),
+        m_distribution(0., 1.) {}
+
+  template<typename B, typename E, typename G>
+  template<typename BeginFwd, typename EndFwd, typename GranularityFwd>
+  Range<B, E, G>::Range(BeginFwd&& begin, EndFwd&& end,
+      GranularityFwd&& granularity, Interval interval)
+      : m_begin(std::forward<BeginFwd>(begin)),
+        m_end(std::forward<EndFwd>(end)),
+        m_round([granularity = std::forward<GranularityFwd(granularity)](
+            const Type& value){
+          auto flr = granularity * static_cast<int64_t>(value / granularity);
+          auto ceil = granularity * (static_cast<int64_t>(value /
+              granularity) + 1);
+          return std::abs(value - flr) <= std::abs(ceil - value) ?
+            flr : ceil;
+        }),
+        m_interval(interval),
+        m_engine(std::random_device()()),
+        m_distribution(0., 1.) {}
+
+  template<typename B, typename E, typename G>
+  typename Range<B, E, G>::Type Range<B, E, G>::generate(
+      Evaluator& evaluator) {
+    auto begin = evaluator.evaluate(m_begin);
+    auto end = evaluator.evaluate(m_end);
     while(true) {
-      auto begin = m_begin.generate(session);
-      auto end = m_end.generate(session);
-      Details::check_range(begin, end);
-      auto distribution = Distribution(begin, end);
-      auto result = distribution(m_engine);
+      auto random_fraction = m_distribution(m_engine);
+      auto distance = static_cast<Type>((end - begin) * random_fraction);
```

Comment:
Due to how operator overloading works with custom types, perform the operation as follows:

```random_fraction * (end - begin)```

The ```static_cast``` should be removed if possible because that involves creating yet another copy of the object. See if you can defer any conversion from any possible intermediate type to ```Type``` until the actual ```return``` which will implicitly perform the conversion only if necessary. Otherwise the ```round``` function can perform the conversion if needed.

A type ```T``` will by default implement ```double * T``` but not necessarily ```T * double```.

Reply:
Here distance represents the offset starting from begin. If I keep it as a double, then value = begin + distance will get promoted to a double, yet the value should be in the return type prior to the rounding. Type casting can indeed be done in the rounding function, but it will be counter-intuitive as it has another function (and I will have to put that line of code in two places rather than one).
With regards to the type casting, there is another issue though. random_fraction from 0 to 1 will not produce a fair probability for the maximum value if the returned type is more narrow than double. E.g. for Range(0, 10, CLOSED), the only way 10 can be produced is when the real distribution returns 1.0 exactly - virtually zero probability. I am not sure what is the best way to fix this issue because each imposes even more requirements on generator arithmetics. If granularity is provided, I can figure out how many possible values are there between begin and end and re-model based on that - but it is possible to use Range for types that do not support granularity.

Reply:
As for implicit conversions, typically narrowing conversions (such as from double to Generator::Type, which may be int or anything) will produce high-order compiler warnings and it is almost universally recommended to cast explicitly in such cases. (You typically have to disable the warning to let the compilation go through.)